mdsa/algorithms/nystrom.py

```diff
@@ -0,0 +1,405 @@
+MDS (LMDS). All three algorithms are based on the Nystrom
+approximation of the eigenvectors and eigenvalues of a matrix. LMDS is
+applies the basic Nystrom approximation, while FastMap and MetricMap
+use generalizations of Nystrom, including deflation and using more
+points to establish an embedding. Empirical experiments on the Reuters
+and Corel Image Features data sets show that the basic Nystrom
+approximation outperforms these generalizations: LMDS is more accurate
+than FastMap and MetricMap with roughly the same computation and can
+become even more accurate if allowed to be slower.``
+
+
+Assume a full distance matrix D for N objects:
+
+        /  E  |  F \
+    D = |-----|----|
+        \ F.t |  G /
+
+
+The correspondong association matrix or centered inner-product
+matrix K is:
+
+        /  A  |  B \
+    K = |-----|----|
+        \ B.t |  C /
+
+where A and B are computed as follows
+
+    A_ij = - 0.50 * (E_ij^2 -
+                 1/m SUM_p E_pj^2 -
+                 1/m SUM_q E_iq^2 +
+                 1/m^2 SUM_q E_pq^2
+
+B is computed as in Landmark MDS, because it's simpler and works
+better according to Platt):
+
+    B_ij = - 0.50 * (F_ij^2 - 1/m SUM_q E_iq^2)
+
+
+In order to approximate an MDS mapping for full matrix you only need E
+and F from D as seed matrix. This will mimick the distances for m seed
+objects. E is of dimension m x m and F of m x (N-m)
+
+E and F are then used to approximate and MDS solution x for the full
+distance matrix:
+
+
+x_ij =  sqrt(g_j) * U_ij, if i<=m
+        and
+        SUM_p B_pi U_pj / sqrt(g_j)
+
+where U_ij is the i'th component of the jth eigenvector of A (see
+below) and g_j is the j'th eigenvalue of A.  The index j only runs
+from 1 to k in order to make a k dimensional embedding.
+"""
+class Nystrom(Algorithm):
+    def __init__(self):
+        super(Nystrom, self).__init__(algorithm_name='nystrom')
+
+    def run(self, distance_matrix, num_dimensions_out=10):
+        n_seeds = int(log(distance_matrix.shape[0], 2)) ** 2
+        if n_seeds <= num_dimensions_out:
+            n_seeds = num_dimensions_out + 1
+        elif n_seeds >= distance_matrix.shape[0]:
+            n_seeds = distance_matrix.shape[0] - 1
+
+        sample_distmtx = array(sample(distance_matrix, n_seeds))
+
+        eigenvectors = self._nystrom(sample_distmtx, num_dimensions_out)
+
+        eigenvalues = [float('nan')] * num_dimensions_out
+        percentages = [float('nan')] * num_dimensions_out
+        return array(eigenvectors), array(eigenvalues), array(percentages)
+
+    def _nystrom(self, seed_distmat, dimensions, PRINT_TIMINGS=False):
+        """Computes an approximate MDS mapping of an (unknown) full distance
+        matrix using a kxn seed distance matrix.
+
+        Returned matrix has the shape seed_distmat.shape[1] x dim
+        """
+
+        if not seed_distmat.shape[0] < seed_distmat.shape[1]:
+            raise ValueError("seed distance matrix should"
```

Comment:
I believe this will print out as "...shouldhave...", possible to add a space after should?